update spec and fix code snippets (#9334)
cosmos/cosmos-sdk: x/authz/spec/03_messages.md

# Messages

In this section we describe the Msg Protobuf service for the authz module.

> ## Grant

Author: I wanted to discuss and describe service methods, not the request objects. That's why I'm not using `Msg` prefix.

> ## Grant

Reviewer: The title of this document is `Messages`. The next line of this section is "An authorization-grant is created using the `MsgGrant` message." Then we provide a code snippet for `MsgGrgant`. And then we further discuss the message, "This message is expected to fail if..." Switching to the method names as the titles for each section without changing the content or even mentioning the method is confusing.

> ## Grant

Author: OK, I can rename the file and the header to `03_tx_service`.

> ## Grant

Author: Each service has exactly one argument, so it makes sense to describe the argument (`MsgGrant`) in this case.

An authorization-grant is created using the `MsgGrant` message.

> +++ https://github.com/cosmos/cosmos-sdk/blob/v0.43.0-beta1/proto/cosmos/authz/v1beta1/tx.proto#L33-L37

Author: ditto

> +++ https://github.com/cosmos/cosmos-sdk/blob/v0.43.0-beta1/proto/cosmos/authz/v1beta1/tx.proto#L33-L37

Reviewer: Not sure I understand "ditto" here. The declaration starts on line 32. This still needs to be fixed.

> +++ https://github.com/cosmos/cosmos-sdk/blob/v0.43.0-beta1/proto/cosmos/authz/v1beta1/tx.proto#L33-L37

Author: yes, fixed it.

This message is expected to fail if:

- both granter & grantee have same address.
- provided `Expiration` time less than current unix timestamp.
- provided `Authorization` is not implemented.
-
- Authorization Method doesn't exist (there is no defined handler in the app router to handle that Msg types)

> If there is already a grant for `(granter, grantee, AuthorizationType)` triple, then the new grant will overwrite the previous one. To update or extend an existing grant a new grant with same `(granter, grantee, AuthorizationType)` should be created.

Reviewer: Also, see [comment](https://github.com/cosmos/cosmos-sdk/pull/9334#issuecomment-841708265).

## Revoke

An allowed authorization can be removed with `MsgRevoke` message.

> +++ https://github.com/cosmos/cosmos-sdk/blob/v0.43.0-beta1/proto/cosmos/authz/v1beta1/tx.proto#L60-L64

Reviewer: ```suggestion
+++ https://github.com/cosmos/cosmos-sdk/blob/v0.43.0-beta1/proto/cosmos/authz/v1beta1/tx.proto#L58-L64
```

Reviewer: This does not need to be changed. Maybe this is where you meant to respond with "ditto". Resolving the suggestion.

This message is expected to fail if:

- both granter & grantee have same address.
- provided `MsgTypeUrl` is empty.

> ## Exec

Author: ditto, let's continue discussion in one thread.

In this section we describe the processing of messages for the authz module.

## MsgGrant

An authorization grant is created using the `MsgGrant` message.
If there is already a grant for the `(granter, grantee, Authorization)` triple, then the new grant will overwrite the previous one. To update or extend an existing grant, a new grant with the same `(granter, grantee, Authorization)` triple should be created.

+++ https://github.com/cosmos/cosmos-sdk/blob/v0.43.0-beta1/proto/cosmos/authz/v1beta1/tx.proto#L32-L37

> The message handling should to fail if:

Reviewer 2: ```suggestion
The message handling should fail if:
```

- both granter and grantee have the same address.
- provided `Expiration` time is less than current unix timestamp.
- provided `Grant.Authorization` is not implemented.
- `Authorization.MsgTypeURL()` is not defined in the router (there is no defined handler in the app router to handle that Msg types).

## MsgRevoke

> A grant can be removed with `MsgRevoke` message.

Reviewer 2: ```suggestion
A grant can be removed with the `MsgRevoke` message.
```

+++ https://github.com/cosmos/cosmos-sdk/blob/v0.43.0-beta1/proto/cosmos/authz/v1beta1/tx.proto#L60-L64

The message handling should fail if:

- both granter and grantee have the same address.
- provided `MsgTypeUrl` is empty.

> NOTE: `MsgExec` message removes a grant if the grant exhausted.

Reviewer 2: ```suggestion
NOTE: The `MsgExec` message removes a grant if the grant has expired.
```

## MsgExec

> When a grantee wants to execute transaction on behalf of a granter, it must send `MsgExec`.

Reviewer 2: ```suggestion
When a grantee wants to execute a transaction on behalf of a granter, they must send `MsgExec`.
```

+++ https://github.com/cosmos/cosmos-sdk/blob/v0.43.0-beta1/proto/cosmos/authz/v1beta1/tx.proto#L47-L53

> The method should fail if:

Reviewer 2: ```suggestion
The message handling should fail if:
```
